lib/functions/versions.bash

```diff
@@ -165,36 +165,73 @@ latest_all() {
     for plugin_path in "$plugins_path"/*/; do
       plugin_name=$(basename "$plugin_path")
 
+      local all_versions
+      all_versions=$(list_all_command "$plugin_name")
+
+      local installed_versions
+      installed_versions=$(list_installed_versions "$plugin_name")
+
+      # Get a distinct list of distributions/variants
+      local installed_variants
+      installed_variants=$(printf "$installed_versions" |
+        sed -e "s/\(^.*\)\-[0-9].*/\1/" |   # extract variant prefix
+        sort -u |                           # unique
+        sed -e "s/^[[:space:]]*[0-9].*//" | # remove versions without variant prefix
+        sed '/^[[:space:]]*$/d')            # remove empty lines
```

Comment:
I'm uncomfortable with this being part of asdf. I know we have similar logic that already exists in a few places, but this doesn't feel like a future-proof solution. We're effectively using versions to store more than the version number. Versions are a combination of tool type (e.g. openjdk) and version (e.g. 1.2.3). I know this is a pretty fundamental change, but I think we might be better off either introducing a type field in addition to version, or specifying that plugins only support one type of thing (e.g. instead of a java plugin we have an openjdk plugin). Thoughts?

Reply:
>specifying that plugins only support one type of thing
interesting proposal. I know this would also upset the Hashicorp plugin which manages the binaries for the Hashi ecosystem and not just different sources of the same runtime. I feel like Java and Hashicorp plugins are different enough and both valid in the spec we have defined 🤔